### ISSUE TITLE
Different behaviour between `Include` and `ThenInclude` when querying self-referencing entities

### ISSUE THREAD

**Issue by adriancentrocol** [dotnet/efcore#35262]
## When using the `ThenInclude` method to include self-referencing list of children items, how can I prevent it from returning both parent and child items at the top level of the list?

**Expected Behavior:** The `SubItems` should only be included within their respective parent `Items` and not at the top level of the query result.  
**Actual Behavior:** The `SubItems` are included at the top level of the query result as well as within their respective parent `Items`.

### Example code

```C#
public class Booking
{
    public Guid Id { get; set; }
    public List<LineItem> LineItems { get; set; }
}

public class LineItem
{
    public Guid Id { get; set; }
    public Guid BookingId { get; set; }
    public LineItem ParentLineItem { get; set; }
    public Guid? ParentLineItemId { get; set; }
    public List<LineItem> SubLineItems { get; set; }
}

public class MyDbContext : DbContext
{
    public DbSet<Booking> Bookings{ get; set; }
    public DbSet<LineItem> LineItems{ get; set; }

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<Booking>(entity =>
        {
            entity.HasKey(e => e.Id);

            entity.HasMany(e => e.LineItems)
                  .WithOne()
                  .HasForeignKey(e => e.BookingId);
        });

        modelBuilder.Entity<LineItem>(entity =>
        {
            entity.HasKey(e => e.Id);

            entity.HasMany(e => e.SubLineItems)
                  .WithOne(e => e.ParentLineItem)
                  .HasForeignKey(e => e.ParentLineItemId);
        });
    }
}
```

In the example below, the result's `LineItems` list includes both parent and sub line items at the top level, as well as sub line items within each parent.
```C#
var result = await _context.Bookings
    .Include(b => b.LineItems
        .Where(l => !l.ParentLineItemId.HasValue))
        .ThenInclude(l => l.SubLineItems)
    .ToListAsync();
```

In the example below, the result is a list of `LineItems` which only includes parent line items at the top level, and the sub line items are properly nested within their parents.
```C#
var result = await _context.LineItem
    .Where(l => !l.ParentLineItemId.HasValue)
    .Include(l => l.SubLineItems)
    .ToListAsync();
```

How can I achieve the same result when using a `ThenInclude`?

### Provider and version information

EF Core version: `7.0.18`
Database provider: `Microsoft.EntityFrameworkCore.SqlServer`
Target framework: `.NET 7.0`

**Comment by ajcvickers** [comment]
@adriancentrocol In the first query, two relationships are loaded: `LineItem` instances related to the `Booking` and also any sub `LineItem` instances related to the `LineItem`s loaded for the booking.

In the second query, only the relationship between the given `LineItem` and its subitems is loaded. The relationship between `LineItem` and `Booking` is not loaded.

See [Relationships](https://learn.microsoft.com/en-us/ef/core/modeling/relationships) for more info.

**Comment by adriancentrocol** [comment]
Hi @ajcvickers,

Thanks for the response. I think there might be a misunderstanding regarding the issue I'm having. To clarify, here's an example of the results I get from both examples in a JSON format:

**Querying at the Booking level**
```c#
var result = await _context.Bookings
    .Include(b => b.LineItems
        .Where(l => !l.ParentLineItemId.HasValue))
        .ThenInclude(l => l.SubLineItems)
    .ToListAsync();
```
```json
{
  "id": "booking-id",
  "lineItems": [
    {
      "id": "line-item-id",
      "parentLineItemId": null,
      "subLineItems": [
        {
          "id": "sub-line-item-id",
          "parentLineItemId": "line-item-id",
          "subLineItems": []
        }
      ]
    },
    {
      "id": "sub-line-item-id",
      "parentLineItemId": "line-item-id",
      "subLineItems": []
    },
  ]
}
```

**Querying at the LineItem level (this is the result I was expecting to see for the previous query)**
```c#
var result = await _context.LineItem
    .Where(l => !l.ParentLineItemId.HasValue)
    .Include(l => l.SubLineItems)
    .ToListAsync();
```
```json
[
  {
    "id": "line-item-id",
    "parentLineItemId": null,
    "subLineItems": [
      {
        "id": "sub-line-item-id",
        "parentLineItemId": "line-item-id",
        "subLineItems": []
      }
    ]
  }
]
```

In the actual output, child items appear both at the top level of the list and nested within their respective parent items. In the expected output, child items should only appear nested within their parent items.

Could you please advise on how to adjust the query to achieve the expected result, or let me know if this is a bug or a limitation of the current implementation?

Thank you!

**Comment by adriancentrocol** [comment]
I have resolved the issue. It turns out that the behavior was due to each `LineItem` having a `BookingId` property and the defined relationship between `LineItems` and `Bookings`. Both parent line items and sub line items had the `BookingId` set, which caused EF Core to use that relationship to pull the sub line items into the top-level list when querying at the `Booking` level.

The solution was to set the `BookingId` column to null on the sub line items to ensure that their only relationship is to their parent line item.
